instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb

```diff
@@ -0,0 +1,87 @@
+                span_name = "#{otel_config[:job_class_span_names] ? job.class : job.queue_name} send"
+                span_attributes = job_attributes(job)
+                otel_tracer.in_span(span_name, attributes: span_attributes, kind: span_kind) do
+                  OpenTelemetry.propagation.inject(job.metadata)
+                  block.call
+                end
+              end
+
+              around_perform do |job, block|
+                span_kind = job.class.queue_adapter_name == 'inline' ? :server : :consumer
+                span_name = "#{otel_config[:job_class_span_names] ? job.class : job.queue_name} process"
+                span_attributes = job_attributes(job).merge('messaging.operation' => 'process')
+
+                propagated_context = OpenTelemetry.propagation.extract(job.metadata)
+
+                if propagation_style == :link
+                  span_links = [
+                    OpenTelemetry::Trace::Link.new(OpenTelemetry::Trace.current_span(propagated_context).context)
+                  ]
+                end
+
+                parent_context = (propagation_style == :child ? propagated_context : OpenTelemetry::Context.current)
+                OpenTelemetry::Context.with_current(parent_context) do
+                  otel_tracer.in_span(span_name, attributes: span_attributes, links: span_links, kind: span_kind) do |span|
+                    span.set_attribute('messaging.active_job.executions', job.executions)
+                    block.call
+                  end
+                end
+              ensure
+                # We may be in a job system (eg: resque) that forks and kills worker processes often.
+                # We don't want to lose spans by not flushing any span processors, so we force it here.
+                otel_tracer.tracer_provider.force_flush
+              end
+            end
+          end
+
+          private
+
+          def job_attributes(job)
+            otel_attributes = {
+              'messaging.destination_kind' => 'queue',
+              'messaging.system' => job.class.queue_adapter_name,
+              'messaging.destination' => job.queue_name,
+              'messaging.message_id' => job.job_id,
+              'messaging.active_job.scheduled_at' => job.scheduled_at,
+              'messaging.active_job.priority' => job.priority
+            }
+
+            otel_attributes['net.transport'] = 'inproc' if %w[async inline].include?(job.class.queue_adapter_name)
+
+            otel_attributes.compact
+          end
+
+          def otel_tracer
+            ActiveJob::Instrumentation.instance.tracer
+          end
+
+          def otel_config
+            ActiveJob::Instrumentation.instance.config
+          end
+
+          def propagation_style
+            # :link is the default option, but we put the fall-back here explicitly
+            # mainly to make testing things easier.
+            otel_config[:propagation_style] || :link
```

Comment:
I had trouble re-installing the instrumentation after changing config options in test. But ... really the truth is that I'm actually just bad at writing tests. I see how you did it in the resque PR and I'll adapt that method! 😄

Reply:
I updated things in aca3d053 based on reading the resque instrumentation tests. I've found that the test suite seems to fail if I re-install over and over again, which I suspect is because ActiveJob doesn't like getting `patch`ed over and over again (suddenly tests which expect 3 spans start seeing ... a lot of spans). I don't have solid evidence for that, though. Lemme know what you think about the approach in aca3d053 and if that's any better/worse than what came before. 😆

Reply:
That makes sense, I think the approach taken is reasonable one given multiple patches don't behave nicely.